### PR TITLE
Adding SQLiteDsn Type

### LINKED ISSUE #7515: add SqliteDsn field type
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

We already have support for [`PostgresDsn`, `MariaDBDsn` and other similar field types](https://docs.pydantic.dev/latest/usage/types/urls/), it would be useful to also have an `SqliteDsn` field type.

For example, an application using SQLAlchemy and supporting different database backends could validate the DSN like this:

```python
class Settings(BaseSettings):
    db_dsn: Union[PostgresDsn, MariaDBDsn, SqliteDsn]
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [X] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by sydney-runkle**
## Change Summary

Adding support for `SQLiteDsn` type

## Related issue number

Fix #7515 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [ ] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @adriangb

### CODE REVIEW

File: pydantic/networks.py [hunk 1]
```diff
@@ -154,6 +156,10 @@ def __hash__(self) -> int:
     ),
 ]
 """A type that will accept any MariaDB DSN."""
+SQLiteDsn = Annotated[
+    FilePath,
+    StringConstraints(pattern=r'.*\.db'),
+]
```
Comment (sydney-runkle) [inline pydantic/networks.py:162]: Does this make sense? As far as I understand, you just need a file name connect to a SQLite database, so this case doesn't mirror the other `Dsn` types we support.

File: tests/test_networks.py [hunk 2]
```diff
@@ -417,6 +418,15 @@ class Model(BaseModel):
...
 
+def test_sqlite_dsns() -> None:
+    class Model(BaseModel):
+        a: SQLiteDsn
+
+    example_db_path: str = 'tests/mocks/test_db.db'
+
+    assert str(Model(a=example_db_path).a) == example_db_path
+
+
```
Comment (sydney-runkle) [inline tests/test_networks.py:429]: Happy to change how we're testing this if we don't want to include a dummy db file in a new `mocks` folder.

**Comment by sydney-runkle** [comment]
please review

**Comment by sydney-runkle** [comment]
Currently working on fixing tests + adding new test cases...

**Review by adriangb — CHANGES_REQUESTED** [review]

File: pydantic/networks.py [hunk 3]
```diff
@@ -160,9 +160,12 @@ def __hash__(self) -> int:
...
 
-def _sqlite_dsn_regex_validator(path: Path) -> str:
-    assert re.match(pattern=r'.*\.db', string=str(path)), 'SQLiteDsn file name must end in `.db`'
-    return str(path)
+def _sqlite_dsn_regex_validator(path: Path) -> Path:
+    try:
+        assert re.match(pattern=r'.*\.db', string=str(path))
+        return path
+    except AssertionError:
+        raise PydanticCustomError('file_type_not_valid', 'SQLiteDsn file name must end in `.db`')
```
Comment (adriangb) [inline pydantic/networks.py:168]: Why not `if not re.match(...):`?

Comment (sydney-runkle) [reply pydantic/networks.py:168]: No reason - changed, thanks!

File: tests/mocks/test_not_db
Comment (adriangb) [inline tests/mocks/test_not_db:1]: I’m confused as to why we need these. It’s not validating that the file actually exists, right?

**Review by adriangb — CHANGES_REQUESTED** [review]
Can you provide a source on why the file extension is what we are validating? How standard is this file extension? Does SQLite just not work if it’s anything else?

File: tests/mocks/test_not_db
Comment (sydney-runkle) [reply tests/mocks/test_not_db:1]: Defining the type as an annotation of `FilePath` does validate that the file actually exists - is this not behavior that we want to include?

**Comment by sydney-runkle** [comment]
> Can you provide a source on why the file extension is what we are validating? How standard is this file extension? Does SQLite just not work if it’s anything else?

Yeah, I think I misunderstood. These [sqlite docs](https://www.sqlite.org/shortnames.html) seem to suggest that other file types might work.

**Comment by sydney-runkle** [comment]
> > Can you provide a source on why the file extension is what we are validating? How standard is this file extension? Does SQLite just not work if it’s anything else?
> 
> Yeah, I think I misunderstood. These [sqlite docs](https://www.sqlite.org/shortnames.html) seem to suggest that other file types might work.

[More confirmation of this](https://www.sqlite.org/appfileformat.html#:~:text=Single%2DFile%20Documents.&text=SQLite%20does%20not%20have%20any,%22belonging%22%20to%20the%20application.)

File: tests/mocks/test_not_db
Comment (adriangb) [reply tests/mocks/test_not_db:1]: Hmm I’m not sure. I can see a scenario where you’re just proxying some data from one place to another and don’t necessarily care if the files exist or not. I’m inclined to have the options for both. Maybe `SQLiteFile` and `ConcreteSQLiteFile` or similar. @samueldeklund wdyt?

**Comment by adriangb** [comment]
If SQLite works with any file extension I feel that this is too under specified for us to create a type for it. Do we have a general “any file with this given extension” type? I imagine it’d have to use Annotated.

**Comment by sydney-runkle** [comment]
> If SQLite works with any file extension I feel that this is too under specified for us to create a type for it. Do we have a general “any file with this given extension” type? I imagine it’d have to use Annotated.

We can easily annotate `Path` or `FilePath` with custom validators + get this effect, so I'd agree that it's too underspecified to merit a new type. Closing...

File: tests/mocks/test_not_db
Comment (samueldeklund) [reply tests/mocks/test_not_db:1]: > Hmm I’m not sure. I can see a scenario where you’re just proxying some data from one place to another and don’t necessarily care if the files exist or not. I’m inclined to have the options for both. Maybe `SQLiteFile` and `ConcreteSQLiteFile` or similar. @samueldeklund wdyt?

I assume you meant to tag @samuelcolvin?

Comment (adriangb) [reply tests/mocks/test_not_db:1]: Yes apologies

**Pull request closed without merging**